library/ziti.c

```diff
@@ -2137,28 +2137,38 @@ static void api_session_cb(ziti_api_session *api_sess, const ziti_error *err, vo
             goto done;
         }
 
-        struct tm exp;
-        ztx->id_creds.cert->get_expiration(ztx->id_creds.cert, &exp);
-        time_t now = time(0);
-        time_t exptime = mktime(&exp);
-
-        bool renew = exptime - now < ztx->opts.cert_extension_window * ONE_DAY;
-        if (!renew) {
+        if (!api_sess->is_cert_extendable) {
+            ZTX_LOG(DEBUG, "identity certificate is not renewable");
             goto done;
         }
 
-        if (!api_sess->is_cert_extendable) {
-            ZTX_LOG(WARN, "identity certificate is not renewable");
-            goto done;
+        struct tm exp;
+        if (api_sess->cert_extend_requested || api_sess->key_roll_requested) {
+            ZTX_LOG(INFO, "controller requested certificate renewal (%s key roll)",
+                    api_sess->key_roll_requested ? "with" : "without");
+        } else {
+            ztx->id_creds.cert->get_expiration(ztx->id_creds.cert, &exp);
+            time_t now = time(0);
+            time_t exptime = mktime(&exp);
+
+            bool renew = exptime - now < ztx->opts.cert_extension_window * ONE_DAY;
+            if (!renew) {
+                goto done;
+            }
+            ZTX_LOG(INFO, "renewing identity certificate exp[%04d-%02d-%02d %02d:%02d]",
+                    1900 + exp.tm_year, exp.tm_mon + 1, exp.tm_mday, exp.tm_hour, exp.tm_min);
         }
 
         if ((ztx->opts.events & ZitiConfigEvent) == 0) {
-            ZTX_LOG(WARN, "identity certificate needs to be renewed but application is not handling ZitiConfigEvent");
+            ZTX_LOG(WARN, "identity certificate needs to be renewed "
+                          "but application is not handling ZitiConfigEvent");
             goto done;
         }
 
-        ZTX_LOG(INFO, "renewing identity certificate exp[%04d-%02d-%02d %02d:%02d]",
-                1900 + exp.tm_year, exp.tm_mon + 1, exp.tm_mday, exp.tm_hour, exp.tm_min);
+        if (api_sess->key_roll_requested) {
+            ZTX_LOG(WARN, "key roll requested, but not yet supported");
+        }
+
```

Comment:
from a debugging perspective, i think there should be a marker log before this section and before the 'done' section below. It wasn't obvious to me that just getting here ends up with a new cert.

Reply:
it is more of a TODO. this shouldn't be here long